rpcping ncreatef

mattbenjamin opened the pull request:
I might be missing something, but it looked to me like the trick to talking to nfs-ganesha is to bypass the binder more-or-less as an nfsv4 backchannel does.

I also saw some odd behavior on some runs, lots to look at, but it appears to work mostly:

17:27:26.628457 IP localhost.60114 > localhost.nfs: Flags [P.], seq 13157:13201, ack 8373, win 342, options [nop,nop,TS val 4248668763 ecr 4248668763], length 44: NFS request xid 300 40 null
17:27:26.628508 IP localhost.nfs > localhost.60114: Flags [P.], seq 8373:8401, ack 13201, win 555, options [nop,nop,TS val 4248668763 ecr 4248668763], length 28: NFS reply xid 300 reply ok 24 null
17:27:26.631487 IP localhost.nfs > localhost.60116: Flags [P.], seq 8373:8401, ack 13201, win 551, options [nop,nop,TS val 4248668766 ecr 4248668763], length 28: NFS reply xid 300 reply ok 24 null

I look forward to all the flames.

mattbenjamin committed:
rpcping: permit client setup by connected socket (omitting bind)

Hook old and new opts via getopt_long.  The proto and host (dest)
arguments remain positional.  Option names from review.

Looks like this does the expected with:
rpcping TCP lemon --nobind -p 2049 -t 1 -c 300 -m 100003 -v 4 -s 0
Signed-off-by: Matt Benjamin <mbenjamin@redhat.com>

was4 commented:
On 3/9/18 5:38 PM, Matt Benjamin wrote:
> I might be missing something, but it looked to me like the trick to talking to nfs-ganesha is to bypass the binder more-or-less as an nfsv4 backchannel does.
> 
I'm not sure this is a good idea, unless we are really desperate for
Ganesha numbers.  I've already shown that we get 85,000 per second in
ntirpc itself, which should be more than enough for now.

The backchannel works because we already have a forward channel.

Let's figure out why rpcbind isn't working on our systems.  Or SE
Linux isn't allowing contact.  Or whatever.

Needs a lot better debugging output than a single message number for
all such problems.  There are sub-message numbers in rpcbind and
getnetconfig and getnetpath, but they aren't being reported.  I've
looked into replacing their annoying thread-local-variables.


> I look forward to all the flames.
> 
This doesn't handle IPv6, unlike my previous commit.  That needs to be
fixed before this is approved.  Traditionally (that is ping and
traceroute), we'd have -4 and -6 for 4-only and 6-only respectively.

Presumably we will handle raw interface via rpcbind, as existing.  But
the Usage information is incorrect.


> 
>         You can view, comment on, or merge this pull request online at:
> 
> https://github.com/nfs-ganesha/ntirpc/pull/115
> 
> 
>         Commit Summary
> 
>   * rpcping: permit binding by connected socket (omitting bind)
>   * rpcping: hook old and new opts via getopt_long_only
> 
Not a big fan of making every parameter require a - flag.  Prefer
required parameters to be positional.  Obviously the original coder
agreed with me on that point!

Protocol is always required, and makes sense to be the first
positional parameter.

rpcping <raw|rdma|tcp|udp> [<destination>]

Because for raw, destination is unused, it should be the last
positional parameter.

If you really must have a flag, traditionally -P for --protocol.

It does make it easier to change some optional parameters without
regard to their order.

You are missing process.  Let's use -s for --process <process>,
-m for --program <program>.  I'd already specified defaults.

Also -b for --rpcbind -- you have a boolean rpcbind flag, but
nothing to set it yet.

Traditional -p for --port, -c for --count (as-is).

That leaves -t for --threads (as-is), -v for --version (as-is).
I've already specified defaults.

I'm sure there is some package out there that handles "-" and "--.
Use the same as traceroute?

mattbenjamin commented:
It is a good idea, because NFSv4 and above -do not use the binder-.  The whole binder integration theme is a red herring.  Don't care about the other details.

mattbenjamin commented:
The whole point of the project is to support hfs-ganesha, so yeah, we are "desperate for" ganesha numbers.

mattbenjamin commented:
My use of getopt was just a swag, I don't think the options to this ping program really justify a large effort, or sourcing an option package other than getopt.  getopt_long handles - and --, but - can either be an alias for -- (long_only) or the short opt, I think.  It should be more than sufficient.

mattbenjamin commented:
Supports ipv6.  Options changed to match the above.

was4 committed:
rpcping.c fix indent, option spelling, defaults

Traditionally spell out options, not variable names.
Defaults should match NFS version 3, lowest common denominator.

was4 reviewed (approved):
Don't know why you are removing tigran credit? Fixed some other bits. Presumably we'll find any other problems during testing.
